File: static/media/js/karel/karelModel.js
/**
 * Class: KarelModel
 * -----------------
 * The KarelModel class is in charge of storing and updating
 * the underlying representation of Karel and her world. 
 * Supports deep copy.
 */
function KarelModel() {

   var that = {};

   that.dir = Const.KAREL_EAST;
   that.karelRow = 0;
   that.karelCol = 0;
   that.beepers = null;
   that.trays = null;
   that.exits = null;
   that.walls = null;
   that.squareColors = null;
   that.rows = 0;
   that.cols = 0;
   that.bag = 0;

   that.deepCopy = function() {
      var newModel = KarelModel();
      newModel.dir = that.dir;
      newModel.karelRow = that.karelRow;
      newModel.karelCol = that.karelCol;
      newModel.beepers = that.beepers.deepCopy();
      newModel.trays = that.trays.deepCopy();
      newModel.exits = that.exits.deepCopy();
      newModel.walls = that.walls.deepCopy();
      newModel.squareColors = that.squareColors.deepCopy();
      newModel.rows = that.rows;
      newModel.cols = that.cols;
      newModel.bag = that.bag;
      return newModel;
   }

   that.equals = function(otherModel) {
      if (that.dir != otherModel.dir) return false;
      if (that.karelRow != otherModel.karelRow) return false;
      if (that.karelCol != otherModel.karelCol) return false;
      if (!that.beepers.equals(otherModel.beepers)) return false;
      if (!that.trays.equals(otherModel.trays)) return false;
      if (!that.exits.equals(otherModel.exits)) return false;
      if (!that.squareColors.equals(otherModel.squareColors)) return false;
      if (!that.bag != otherModel.bag) return false;
      return true;
   }

    that.exit = function() {
        if(that.exits.exitPresent(that.karelRow, that.karelCol)) {
            return true;
        }
        return false;
    }

    that.move = function() {
        var newRow = that.karelRow;
		var newCol = that.karelCol;
		switch(that.dir) {
			case Const.KAREL_EAST: newCol = newCol + 1; break;
			case Const.KAREL_WEST: newCol = newCol - 1; break;
			case Const.KAREL_NORTH: newRow = newRow - 1; break;
			case Const.KAREL_SOUTH: newRow = newRow + 1; break;
		}
		if(that.walls.isMoveValid(that.karelRow, that.karelCol, newRow, newCol)) {
			that.karelRow = newRow;
			that.karelCol = newCol;
            window.console.log("row: "+newRow+" col: "+newCol);
		} else {
		   error('Front Is Blocked');
           return false;
		}
        return true;
   }

   that.turnLeft = function() {
      var newD = that.dir;
		switch(that.dir) {
            case Const.KAREL_EAST:  newD = Const.KAREL_NORTH; console.log('dir:'+newD); break;
            case Const.KAREL_WEST:  newD = Const.KAREL_SOUTH; console.log('dir:'+newD); break;
            case Const.KAREL_NORTH: newD = Const.KAREL_WEST; console.log('dir:'+newD); break;
            case Const.KAREL_SOUTH: newD = Const.KAREL_EAST; console.log('dir:'+newD); break;	
			default: alert("invalid that.dir: " + that.dir); break;	
		}
		that.dir = newD;
   }

   that.turnRight = function() {
      var newD = that.dir;
		switch(that.dir) {
			case Const.KAREL_EAST:  newD = Const.KAREL_SOUTH; break;
			case Const.KAREL_WEST:  newD = Const.KAREL_NORTH; break;
			case Const.KAREL_NORTH: newD = Const.KAREL_EAST; break;
			case Const.KAREL_SOUTH: newD = Const.KAREL_WEST; break;	
			default: alert("invalid that.dir: " + that.dir); break;	
		}
		that.dir = newD;
   }

   that.pickBeeper = function() {
      if (that.beepers.beeperPresent(that.karelRow, that.karelCol)) {
         that.beepers.pickBeeper(that.karelRow, that.karelCol);
         that.bag++;
      } else {
         error('No Beepers Present');
         return false;
      }
      return true;
   }

    that.putBeeper = function() {
        if (that.bag > 0) {
            that.beepers.putBeeper(that.karelRow, that.karelCol);
            that.bag--;
        } else {
            error('Not carying any beeper');
            return false;
        }
        return true;
    }

    that.putBeeperInTray = function() {
        if (that.bag <= 0) {
            error('Not carying any beeper');
            return false;
        }
        if (that.trays.trayPresent(that.karelRow, that.karelCol)) {
            var numBeepers = that.beepers.numBeepers(that.karelRow, that.karelCol);
            if (that.trays.trayCapacity(that.karelRow, that.karelCol) > numBeepers) {
                that.beepers.putBeeper(that.karelRow, that.karelCol);
            } else {
                error('Tray is full');
                return false;
            }
        } else {
            error('No Tray Present');
            return false;
        }
        return true;
    }

    that.pickBeeperFromTray = function() {
        if (that.trays.trayPresent(that.karelRow, that.karelCol)) {
            var numBeepers = that.beepers.numBeepers(that.karelRow, that.karelCol);
            if (numBeepers > 0) {
                that.beepers.pickBeeper(that.karelRow, that.karelCol);
            } else {
                error('Tray is empty');
                return false;
            }
        } else {
            error('No Tray Present');
            return false;
        }
        return true;
    }

   that.turnAround = function() {
      var newD = that.dir;
		switch(that.dir) {
			case Const.KAREL_EAST:  newD = Const.KAREL_WEST; break;
			case Const.KAREL_WEST:  newD = Const.KAREL_EAST; break;
			case Const.KAREL_NORTH: newD = Const.KAREL_SOUTH; break;
			case Const.KAREL_SOUTH: newD = Const.KAREL_NORTH; break;	
			default: alert("invalid that.dir: " + that.dir); break;	
		}
		that.dir = newD;
   }

   that.paintCorner = function(color) {
      that.squareColors.paintCorner(that.karelRow, that.karelCol, color);
   }

   that.getDirection = function() {
      return that.dir;
   }

   that.getNumRows = function() {
      return that.rows;
   }

   that.getNumCols = function() {
      return that.cols;
   }

   that.getKarelRow = function() {
      return that.karelRow;
   }

   that.getKarelCol = function() {
      return that.karelCol;
   }

   that.getSquareColor = function(r, c) {
      return that.squareColors.getColor(r, c);
   }

   that.getNumBeepers = function(r, c) {
      return that.beepers.getNumBeepers(r, c);
   }

   that.getNumBeepersInTray = function(r, c) {
      if (that.trays.trayPresent(that.karelRow, that.karelCol)) {
          return that.beepers.getNumBeepers(r, c);
      } else {
         error('No Trays Present');
         return false;
      }
      return true;
   }

   that.hasWall = function(r, c) {
      return that.walls.wall(r, c);
   }

	that.beeperPresent = function() {
		return that.beepers.beeperPresent(that.karelRow, that.karelCol);
	}

	that.trayPresent = function() {
		return that.trays.trayPresent(that.karelRow, that.karelCol);
	}

    that.exitPresent = function() {
		return that.exits.exitPresent(that.karelRow, that.karelCol);
    }

    that.trayFull = function() {
        return that.trays.trayCapacity(that.karelRow, that.karelCol) <= that.getNumBeepers(that.karelRow, that.karelCol);
    }

   that.frontIsClear = function() {
      var newRow = that.karelRow;
		var newCol = that.karelCol;
		switch(that.dir) {
			case Const.KAREL_EAST: newCol = newCol + 1; break;
			case Const.KAREL_WEST: newCol = newCol - 1; break;
			case Const.KAREL_NORTH: newRow = newRow - 1; break;
			case Const.KAREL_SOUTH: newRow = newRow + 1; break;
			default: alert("invalid that.dir: " + that.dir); break;		
		}
		var ret = that.walls.isMoveValid(that.karelRow, that.karelCol, newRow, newCol);
        return ret;
	}

	that.rightIsClear = function() {
		var newRow = that.karelRow;
		var newCol = that.karelCol;
		switch(that.dir) {
			case Const.KAREL_EAST: newRow = newRow + 1; break;
			case Const.KAREL_WEST: newRow = newRow - 1; break;
			case Const.KAREL_NORTH: newCol = newCol + 1; break;
			case Const.KAREL_SOUTH: newCol = newCol - 1; break;
			default: alert("invalid that.dir: " + that.dir); break;		
		}
		return that.walls.isMoveValid(that.karelRow, that.karelCol, newRow, newCol);
	}

	that.leftIsClear = function() {
		var newRow = that.karelRow;
		var newCol = that.karelCol;
		switch(that.dir) {
			case Const.KAREL_EAST: newRow = newRow - 1; break;
			case Const.KAREL_WEST: newRow = newRow + 1; break;
			case Const.KAREL_NORTH: newCol = newCol - 1; break;
			case Const.KAREL_SOUTH: newCol = newCol + 1; break;
			default: alert("invalid that.dir: " + that.dir); break;		
		}
		return that.walls.isMoveValid(that.karelRow, that.karelCol, newRow, newCol);
	}

	that.facingNorth = function() {
		return virtualDirection == KAREL_NORTH;	
	}

	that.facingSouth = function() {
		return virtualDirection == KAREL_SOUTH;	
	}

	that.facingEast = function() {
		return virtualDirection == KAREL_EAST;	
	}

	that.facingWest = function() {
		return virtualDirection == KAREL_WEST;	
	}

   that.loadWorld = function(worldText) {
      var lines = worldText.split("\n");

		// get world dimension
		loadDimensionLine(lines[0]);
		
		that.beepers = Beepers(that.rows, that.cols);
        that.trays = Trays(that.rows, that.cols);
        that.exits = Exits(that.rows, that.cols);
		that.walls = Walls(that.rows, that.cols);
      that.squareColors = SquareColors(that.rows, that.cols);

      that.dir = Const.KAREL_EAST;

		// load world details
		for (var i = 1; i < lines.length; i++) {
		   if (lines[i] != '') {
			   loadLine(lines[i]);
		   }
		}
   }

   function extractCoord(coordString) {
      var rParenIndex = coordString.indexOf('(');
      var lParenIndex = coordString.indexOf(')');
      coordString = coordString.substring(rParenIndex + 1, lParenIndex);
      coordStrings = coordString.split(',');
      var row = parseInt(coordStrings[0]);
      var col = parseInt(coordStrings[1]);
      return [row, col];
   }

   function extractCapacity(coordString) {
      var rParenIndex = coordString.indexOf('(');
      var lParenIndex = coordString.indexOf(')');
      coordString = coordString.substring(rParenIndex + 1, lParenIndex);
      coordStrings = coordString.split(',');
      var cap = parseInt(coordStrings[2]);
      return cap;
   }

   function loadDimensionLine(line) {
		var dimensionStrings = line.split(":");

      assert(dimensionStrings[0] == 'Dimension', 'World file malformed');
      var coord = extractCoord(dimensionStrings[1]);

		that.rows = coord[0];
		that.cols = coord[1];
   }

   function error(msg) {
      window.console.log(msg);
   }

   function placeKarel(row, col) {
      that.karelRow = row;
      that.karelCol = col;
   }
   
   function loadWallLine(line) {
       var coord = extractCoord(line);
       var row = coord[1];
       var col = coord[0];
       if (col >= 0 && row >= 0 && col<that.cols && row<that.rows) {
           that.walls.addWall(row, col);
       }
   }

   function loadBeeperLine(line) {
      var coord = extractCoord(line);
      var row = coord[0];
      var col = coord[1];
      that.beepers.putBeeper(row, col)
   }

   function loadTrayLine(line) {
      var coord = extractCoord(line);
      var row = coord[0];
      var col = coord[1];
      var capacity = extractCapacity(line);
      that.trays.putTray(row, col, capacity)
   }

   function loadExitLine(line) {
       var coord = extractCoord(line);
       var row = coord[0];
       var col = coord[1];
       that.exits.putExit(row, col)
   }

   function loadKarelLine(line) {
      var coord = extractCoord(line);
      var row = coord[0];
      var col = coord[1];
      placeKarel(row, col);
      if (line.indexOf('west') != -1) {
         that.dir = Const.KAREL_WEST;
      }
      if (line.indexOf('south') != -1) {
         that.dir = Const.KAREL_SOUTH;
      }
      if (line.indexOf('north') != -1) {
         that.dir = Const.KAREL_NORTH;
      }
   }

   function loadLine(line) {
		var elements = line.split(":");
		assert(elements.length == 2, line + 'World file missing :');
		var key = elements[0];

		if (key == "Karel")  {
			loadKarelLine(elements[1]);
		} else if (key == "Wall")  {
			loadWallLine(elements[1]);
		} else if (key == "Beeper") {
			loadBeeperLine(elements[1]);
		} else if (key == "Tray") {
			loadTrayLine(elements[1]);
		} else if (key == "Exit") {
			loadExitLine(elements[1]);
		} 
	}

   return that;

}

File: karel_model.py
import json

from pykarel.karel.beepers import Beepers
from pykarel.karel.karel_constants import KAREL_EAST, KAREL_WEST, KAREL_NORTH, KAREL_SOUTH
from pykarel.karel.walls import Walls


def error(txt):
    print(txt)


def log(txt):
    print(txt)

class KarelEntity:
    def __init__(self, handle, row, col, dir):
        self.handle = handle
        self.row = row
        self.col = col
        self.dir = dir

    def __str__(self):
        return "Karel '{}' at {}, {} facing {}".format(self.handle, self.row, self.col, self.dir) 

class KarelModel:
    def __init__(self):
        self.beepers = None
        self.trays = None
        self.exits = None
        self.walls = None
        self.karels = {}
        self.rows = 0
        self.cols = 0

    def exit(self, handle):
        if self.exits.exit_present(self.karels[handle].row, self.karels[handle].col):
            return True
        return False

    def move(self, handle):
        new_row = self.karels[handle].row
        new_col = self.karels[handle].col
        if self.karels[handle].dir == KAREL_EAST:
            new_col += 1
        elif self.karels[handle].dir == KAREL_WEST:
            new_col -= 1
        elif self.karels[handle].dir == KAREL_NORTH:
            new_row -= 1
        elif self.karels[handle].dir == KAREL_SOUTH:
            new_row += 1
        if self.walls.is_move_valid(self.karels[handle].row, self.karels[handle].col, new_row, new_col):
            self.karels[handle].row = new_row
            self.karels[handle].col = new_col
            log("row: {} col: {}".format(new_row, new_col))
            return True
        else:
            error("Front is blocked")
        return False

    def turn_left(self, handle):
        new_d = self.karels[handle].dir
        if self.karels[handle].dir == KAREL_EAST:
            new_d = KAREL_NORTH
            log("dir: {}".format(new_d))
        elif self.karels[handle].dir == KAREL_WEST:
            new_d = KAREL_SOUTH
            log("dir: {}".format(new_d))
        elif self.karels[handle].dir == KAREL_NORTH:
            new_d = KAREL_WEST
            log("dir: {}".format(new_d))
        elif self.karels[handle].dir == KAREL_SOUTH:
            new_d = KAREL_EAST
            log("dir: {}".format(new_d))
        else:
            error("invalid dir: {}".format(self.dir))
        self.karels[handle].dir = new_d

    def turn_right(self):
        new_d = self.karels[handle].dir
        if self.karels[handle].dir == KAREL_EAST:
            new_d = KAREL_SOUTH
            log("dir: {}".format(new_d))
        elif self.karels[handle].dir == KAREL_WEST:
            new_d = KAREL_NORTH
            log("dir: {}".format(new_d))
        elif self.karels[handle].dir == KAREL_NORTH:
            new_d = KAREL_EAST
            log("dir: {}".format(new_d))
        elif self.karels[handle].dir == KAREL_SOUTH:
            new_d = KAREL_WEST
            log("dir: {}".format(new_d))
        else:
            error("invalid dir: {}".format(self.dir))
        self.karels[handle].dir = new_d

    def get_direction(self, handle):
        return self.karels[handle].dir

    def get_num_rows(self):
        return self.rows

    def get_num_cols(self):
        return self.cols

    def get_karel_row(self, handle):
        return self.karels[handle].row

    def get_karel_col(self, handle):
        return self.karels[handle].col

    def get_num_beepers(self, row, col):
        return self.beepers.get_num_beepers(row, col)

    def has_wall(self, row, col):
        return self.walls.get_wall(row, col)

    def beepers_present(self):
        return self.beepers.beeper_present(self.karel_row, self.karel_col)

    def frontIsClear(self, handle):
        log('frontIsClear')
        new_row = self.karels[handle].row
        new_col = self.karels[handle].col
        if self.karels[handle].dir is KAREL_EAST:
            new_col += 1
        elif self.karels[handle].dir is KAREL_WEST:
            new_col -= 1
        elif self.karels[handle].dir is KAREL_NORTH:
            new_row -= 1
        elif self.karels[handle].dir is KAREL_SOUTH:
            new_row += 1
        else:
            error("invalid dir: {}".format(self.karels[handle].dir))
        ret = self.walls.is_move_valid(self.karels[handle].row, self.karels[handle].col, new_row, new_col)
        return ret

    def load_world(self, world):
        self.rows = world["dimension"][0]
        self.cols = world["dimension"][1]

        self.beepers = Beepers(self.rows, self.cols)
        self.walls = Walls(self.rows, self.cols)

        for beeper in world["beepers"]:
            self.beepers.put_beeper(beeper[0], beeper[1])

        for wall in world["walls"]:
            self.walls.add_wall(wall[0], wall[1])

        for karel in world["karels"]:
            self.karels[karel[0]] = KarelEntity(karel[0], karel[1], karel[2], KAREL_EAST)

    def dump_world(self):
        world = {}
        world["dimension"] = [self.rows, self.cols]
        world["beepers"] = []
        world["walls"] = self.walls.dump_walls()
        world["karels"] = []
        for karel in self.karels.values():
            log("karel: {}".format(karel))
            world["karels"].append([karel.handle, karel.row, karel.col, karel.dir])
        return world


